pkg/policy/repository.go

```diff
@@ -527,12 +548,25 @@ func (p *Repository) DeleteByLabelsLocked(lbls labels.LabelArray) (ruleSlice, ui
 		}
 	}
 
+	if uid := lbls.Get(labels.LabelSourceK8sKeyPrefix + k8sConst.PolicyLabelUID); deleted > 0 && uid == "" {
+		for uid, indexedRules := range p.rulesIndexByK8sUID {
+			newIndex := indexedRules[:0]
+			for _, r := range indexedRules {
+				if !r.Labels.Contains(lbls) {
+					newIndex = append(newIndex, r)
+				}
+			}
+			if len(newIndex) > 0 {
+				p.rulesIndexByK8sUID[uid] = newIndex
+			} else {
+				delete(p.rulesIndexByK8sUID, uid)
+			}
+		}
+	}
+
 	if deleted > 0 {
 		p.BumpRevision()
 		p.rules = new
```

Comment:
Can't the entire diff be put in here like this?

```golang
if uid := lbls.Get(labels.LabelSourceK8sKeyPrefix + k8sConst.PolicyLabelUID); uid != "" {
	indexedRules, found := p.rulesIndexByK8sUID[uid]
	if found {
		newIndex := indexedRules[:0]
		for _, r := range indexedRules {
			if !r.Labels.Contains(lbls) {
				newIndex = append(newIndex, r)
			}
		}
		if len(newIndex) > 0 {
			p.rulesIndexByK8sUID[uid] = newIndex
		} else {
			delete(p.rulesIndexByK8sUID, uid)
		}
	}
}
```

Reply:
Hmm. Are you talking about the full diff of `DeleteByLabelsLocked`?

If so, we would loose out on the early exit with `UID != ""`. We also still need to filter all the "index" values in case there is no UID.

Reply:
> Hmm. Are you talking about the full diff of DeleteByLabelsLocked?

No. Ok so this would be the full diff for the `DeleteByLabelsLocked`:


```golang
func (p *Repository) DeleteByLabelsLocked(lbls labels.LabelArray) (ruleSlice, uint64, int) {
	deleted := 0
	new := p.rules[:0]
	deletedRules := ruleSlice{}

	if uid := lbls.Get(labels.LabelSourceK8sKeyPrefix + k8sConst.PolicyLabelUID); uid != "" {
		indexedRules, found := p.rulesIndexByK8sUID[uid]
		if !found {
			return deletedRules, p.GetRevision(), deleted
		}
		newIndex := indexedRules[:0]
		for _, r := range indexedRules {
			if !r.Labels.Contains(lbls) {
				newIndex = append(newIndex, r)
			}
		}
		// No rules found
		if len(newIndex) == len(indexedRules) {
			return deletedRules, p.GetRevision(), deleted
		} else if len(newIndex) > 0 {
			p.rulesIndexByK8sUID[uid] = newIndex
		} else {
			delete(p.rulesIndexByK8sUID, uid)
		}
	}

	for _, r := range p.rules {
		if !r.Labels.Contains(lbls) {
			new = append(new, r)
		} else {
			deletedRules = append(deletedRules, r)
			deleted++
		}
	}

	if deleted > 0 {
		p.BumpRevision()
		p.rules = new
		metrics.Policy.Sub(float64(deleted))
	}

	return deletedRules, p.GetRevision(), deleted
}
```